### DISCUSSION TITLE
Ref to another json schema doesn`t work

### DISCUSSION THREAD

**Discussion by HalinaPichukha** [networknt/json-schema-validator#1103] [Q&A] [answered]
Hello :)

I want to implement referensing to the child json schema from the parent one.

I have main json schema "parent-validation-schema.json":
{
  "$schema": "https://json-schema.org/draft/2020-12/schema",
  "title": "parentschema",
  "type": "object",
  "properties": {
    "myProperty": {
          "$ref": "./child-validation-schema.json"
      }
   }
}

and the child-validation-schema.json:
{
    "title": "child schema",
    "type": "object",
    "properties": {
        "myProperty": {
            "type": "String"
        }
    }
}

Also I configured JsonSchemaFactory and JsonSchema (it works grait for the single json schema):

 public class JsonSchemaValidator implements EventValidator<JsonNode> {

  private final JsonSchema jsonSchema;

  public JsonSchemaValidator() {
    final var factory = JsonSchemaFactory.getInstance(V202012);
    jsonSchema = factory.getSchema(JsonSchemaValidator.class.getResourceAsStream("/parent-validation-schema.json"));
    jsonSchema.createExecutionContext().getExecutionConfig().setFormatAssertionsEnabled(true);
  }
}

They are located on the same directory.

I`m basing on the test cases in your repo:
https://github.com/networknt/json-schema-validator/blob/master/src/test/resources/schema/main/main.json
https://github.com/networknt/json-schema-validator/blob/master/src/test/resources/schema/common/child.json
https://github.com/networknt/json-schema-validator/blob/master/src/test/resources/schema/common/child2.json

But I get error: **URI is not absolute**.

I case I use ` "$ref": "file:./child-validation-schema.json" ` or ` "$ref": "classpath./child-validation-schema.json" ` I don`t get any errors, but only the main (parent) validation schema works, and the nested (child) is ignored.

Could you please help me with this issue? :)
Thanks a lot!

**Comment by justin-tay** [comment, marked as answer]
A relative IRI `$ref` needs a base IRI to resolve. As you are using an `InputStream` to load your parent schema and there is no `$id` defined there is no base IRI.

Instead of

```java
jsonSchema = factory.getSchema(JsonSchemaValidator.class.getResourceAsStream("/parent-validation-schema.json"));
```

You can try

```java
jsonSchema = factory.getSchema(SchemaLocation.of("classpath:/parent-validation-schema.json"));
```

Also this line doesn't do anything meaningful as you created an execution context and set the config but never stored the reference to it in order to pass it somewhere.

```java
jsonSchema.createExecutionContext().getExecutionConfig().setFormatAssertionsEnabled(true);
```

Typically this is done by passing a lambda as a parameter to validate if you want to customize the execution context for that run.

```java
Set<ValidationMessage> messages = schema.validate(inputData, InputFormat.JSON, executionContext -> {
    executionContext.getExecutionConfig().setFormatAssertionsEnabled(true);
});
```
